### ISSUE TITLE
Queue not working relaliably on Ruby 1.9.2

### ISSUE THREAD

**Issue by fbjork** [mperham/girl_friday#25]
I have a Ruby 1.9.2 app on Heroku with the following queue:

SYNC_FRIENDS_QUEUE = GirlFriday::WorkQueue.new(:sync_friends) do |msg|
  Account.sync_friends(msg[:account_id])
end

I've noticed that sometimes the job is never executed at random. Don't see anything in the logs when it happens. Any ideas?

threadsafe is not enabled.

**Comment by jc00ke** [comment]
What web server are you using? Unicorn or Rainbows! on cedar are the preferred servers.

**Comment by fbjork** [comment]
Thin on Cedar

**Comment by jc00ke** [comment]
Because Thin is an evented server it's not guaranteed to work well with gf. If you can, switch to Unicorn or Rainbows! and you should see normal usage.

**Comment by fbjork** [comment]
Would it work better on Thin if I enabled threadsafe mode? I'm not too keen on moving to Rainbows! for the sole purpose of supporting GirlFriday.

**Comment by jc00ke** [comment]
I don't think so.

**Comment by mperham** [comment]
@fbjork Try it, what's the worst that could happen?  I've also found in practice that Unicorn works well and Rainbows gave me problems.  So there's that completely anecdotal evidence...

**Comment by fbjork** [comment]
Will give Rainbows a shot and let you know my findings.

**Comment by jc00ke** [comment]
I'm using Unicorn/gf on several projects hosted on Heroku & it works very well. Haven't tried Rainbows! but it'd be cool to hear how it does.

**Comment by fbjork** [comment]
Ok, so I switched to Unicorn. I'm assuming threadsafe! doesn't need to be enabled for this to work right?

**Comment by jc00ke** [comment]
Correct

**Comment by fbjork** [comment]
So it seems switching to Unicorn didn't fix the queue not working reliably. Could it be some gems I'm using that's causing problems? After restarting all Unicorns on Heroku the queue works 100% of the time in the beginning, then after a while some enqueued jobs don't get processed. Any ideas?

This is using the in-memory store.

**Comment by jc00ke** [comment]
Hmm, not really. Can we see the code, or could you set up a repo to repro?

**Comment by fbjork** [comment]
Here's a gist of the code:
https://gist.github.com/1218592

I'll see if I can create a repo with a subset of the code, unfortunately I cannot expose the full code base. Let me know if that helps.

**Comment by jc00ke** [comment]
Nothing jumps out at me. A reproducible example would be great.

**Comment by fbjork** [comment]
I'm using these gems:
https://github.com/cjbottaro/thread_storm
https://github.com/grosser/parallel

Wondering if they are exhausting the thread limit on Heroku some how?

**Comment by jc00ke** [comment]
That very well could be the case. Why are you using parallel? Could you get the same thing w/girl_friday's pipeline?

**Comment by fbjork** [comment]
Using parallel to access several APIs in parallel in a controller action synchronously. Open to suggestions to switch to other gems that work better with GirlFriday if it's the problem here.

**Comment by fbjork** [comment]
Just discovered the GirlFriday::Batch method. Will have a look at it and replace Parallel with this and see if the problems go away.

**Comment by jc00ke** [comment]
Sorry, batch is what I meant to suggest, not pipeline. Glad you found it though!

**Comment by fbjork** [comment]
I tried using GirlFriday::Batch on 0.9.5 and got "deadlock detected" as soon the size went over 1. Any ideas?

**Comment by mperham** [comment]
Can you open a new issue with a stack trace, code snippet, more info about your environment, and whatever else info you can provide?  We have tests which run batch with many threads and they all pass.
